yarn-project/end-to-end/src/e2e_rollup_native_asset_contract.test.ts

```diff
@@ -0,0 +1,218 @@
+    const y = point.buffer.subarray(32, 64);
+    return {
+      x: toBigIntBE(x),
+      y: toBigIntBE(y),
+    };
+  };
+
+  const deployContract = async (initialBalance = 0n, owner = { x: 0n, y: 0n }) => {
+    logger(`Deploying L2 contract...`);
+    const deployer = new ContractDeployer(RollupNativeAssetContractAbi, aztecRpcServer);
+    const tx = deployer.deploy(initialBalance, owner).send({
+      portalContract: portalAddress,
+    });
+    const receipt = await tx.getReceipt();
+    contract = new Contract(receipt.contractAddress!, RollupNativeAssetContractAbi, aztecRpcServer);
+    await contract.attach(portalAddress);
+
+    await tx.isMined(0, 0.1);
+    await tx.getReceipt();
+    logger('L2 contract deployed');
+    return contract;
+  };
+
+  /**
+   * Milestone 2 - L2 to L1
+   */
+  it('Milestone 2.3: Exit funds from L2 to L1', async () => {
+    const initialBalance = 987n;
+    const withdrawAmount = 654n;
+    const [owner] = accounts;
+
+    await deployContract(initialBalance, pointToPublicKey(await aztecRpcServer.getAccountPublicKey(owner)));
+    await expectBalance(owner, initialBalance);
+
+    const { request: initRequest } = await publicClient.simulateContract({
+      account,
+      address: getAddress(portalAddress.toString()),
+      abi: RollupNativeAssetAbi,
+      functionName: 'initialize',
+      args: [getAddress(registryAddress.toString()), `0x${contract.address.toString().slice(2)}`],
+    });
+
+    await walletClient.writeContract(initRequest);
+
+    const ethOutAddress = EthAddress.fromString('0x000000000000000000000000000000000000dead');
+
+    const tx = contract.methods
+      .withdraw(
+        withdrawAmount,
+        pointToPublicKey(await aztecRpcServer.getAccountPublicKey(owner)),
+        ethOutAddress.toField().value,
+      )
+      .send({ from: accounts[0] });
+
+    await tx.isMined(0, 0.1);
+    const receipt = await tx.getReceipt();
+    expect(receipt.status).toBe(TxStatus.MINED);
+    await expectBalance(owner, initialBalance - withdrawAmount);
+
+    // 0x00f714ce, selector for "withdraw(uint256,address)"
+    const content = sha256ToField(
+      Buffer.concat([
+        Buffer.from([0x00, 0xf7, 0x14, 0xce]),
+        toBufferBE(withdrawAmount, 32),
+        ethOutAddress.toBuffer32(),
+      ]),
+    );
+    const contractInfo = await node.getContractInfo(contract.address);
+    // Compute the expected hash and see if it is what we saw in the block.
+    const entryKey = sha256ToField(
+      Buffer.concat([
+        contract.address.toBuffer(),
+        fr(1).toBuffer(), // aztec version
+        contractInfo?.portalContractAddress.toBuffer32() ?? Buffer.alloc(32, 0),
+        fr(publicClient.chain.id).toBuffer(), // chain id
+        content.toBuffer(),
+      ]),
+    );
+
+    const blockNumber = await node.getBlockHeight();
+    const blocks = await node.getBlocks(blockNumber, 1);
+    // If this is failing, it is likely because of wrong chain id
```

Comment:
how do you set the right chain id?

Reply:
Mainly that the two domains need to match, e.g., rollup and l1 need to specify the same chainid for messages to hit the expected hashes, if one use 1 and other use 31337 you won't find occurences of the messages correctly.